pkg/controller/async_backup_operations_controller.go

```diff
@@ -0,0 +1,495 @@
+	}
+
+	if loc.Spec.AccessMode == velerov1api.BackupStorageLocationAccessModeReadOnly {
+		log.Infof("Cannot check progress on async Backup operations because backup storage location %s is currently in read-only mode; marking backup PartiallyFailed", loc.Name)
+		backup.Status.Phase = velerov1api.BackupPhasePartiallyFailed
+
+		err := c.updateBackupAndOperationsJSON(ctx, original, backup, nil, &operationsForBackup{errsSinceUpdate: []string{"BSL is read-only"}}, false, false)
+		if err != nil {
+			log.WithError(err).Error("error updating Backup")
+		}
+		return ctrl.Result{}, nil
+	}
+
+	pluginManager := c.newPluginManager(c.logger)
+	defer pluginManager.CleanupClients()
+	backupStore, err := c.backupStoreGetter.Get(loc, pluginManager, c.logger)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "error getting backup store")
+	}
+
+	operations, err := c.getOperationsForBackup(backupStore, backup.Name)
+	if err != nil {
+		err2 := c.updateBackupAndOperationsJSON(ctx, original, backup, backupStore, &operationsForBackup{errsSinceUpdate: []string{err.Error()}}, false, false)
+		if err2 != nil {
+			return ctrl.Result{}, errors.Wrap(err2, "error updating Backup")
+		}
+		return ctrl.Result{}, errors.Wrap(err, "error getting backup operations")
+	}
+	stillInProgress, changes, opsCompleted, opsFailed, errs := getBackupItemOperationProgress(backup, pluginManager, operations.operations)
+	// if len(errs)>0, need to update backup errors and error log
+	operations.errsSinceUpdate = append(operations.errsSinceUpdate, errs...)
+	backup.Status.Errors += len(operations.errsSinceUpdate)
+	asyncCompletionChanges := false
+	if backup.Status.AsyncBackupItemOperationsCompleted != opsCompleted || backup.Status.AsyncBackupItemOperationsFailed != opsFailed {
+		asyncCompletionChanges = true
+		backup.Status.AsyncBackupItemOperationsCompleted = opsCompleted
+		backup.Status.AsyncBackupItemOperationsFailed = opsFailed
+	}
+	if changes {
+		operations.changesSinceUpdate = true
+	}
+
+	// if stillInProgress is false, backup is complete and needs update
+	// if operations.errsSinceUpdate is not empty, then backup phase needs to change to
+	// BackupPhaseWaitingForPluginOperationsPartiallyFailed and needs update
+	// If the only changes are incremental progress, then no write is necessary, progress can remain in memory
+	if !stillInProgress {
+		if len(operations.errsSinceUpdate) > 0 {
+			backup.Status.Phase = velerov1api.BackupPhaseWaitingForPluginOperationsPartiallyFailed
+		}
+		if backup.Status.Phase == velerov1api.BackupPhaseWaitingForPluginOperations {
+			log.Infof("Marking backup %s Completed", backup.Name)
+			backup.Status.Phase = velerov1api.BackupPhaseCompleted
+		} else {
+			log.Infof("Marking backup %s PartiallyFailed", backup.Name)
+			backup.Status.Phase = velerov1api.BackupPhasePartiallyFailed
+		}
+	}
+	err = c.updateBackupAndOperationsJSON(ctx, original, backup, backupStore, operations, asyncCompletionChanges, changes)
```

Comment:
At least for the current situation of Velero for data movement, I think we don't need to frequently persist the async info, we only need to do it when the async operation comes to a terminal status. Reasons:
1. As we discussed, we don't need to sync the backup in another cluster if the backup doesn't go to a terminal status
2. In the same cluster, we don't support restart recovery for many reasons, Velero itself, the BIA/RIA V2 interface & implementation, the data mover
3. Because of 1 and 2, we can say that the in-memory async operation info always exist in the current cluster as well as the on-air SnapshotBackup CR, the combination of these to info should be enough for any further operation

Moreover, frequently writing to backup storage will invoke unnecessary costs.

Of cause, as a general BIA implementation, for some other operations now and in future, the update during progress may be necessary, so probably we make the in-progress update optional here?

Reply:
Look at the implementation of `updateBackupAndOperationsJSON` -- we only upload the change under two circumstances:
1) The backup is completed
2) A plugin operation errors out.

Under normal operation, where periodic updates just end up changing progress numbers, partial completion, none of that  gets uploaded until either the backup is complete or a user explicitly requests it. So in that sense, it's already optional. We basically delay this until we absolutely have to make the update.

Reply:
OK, thanks. For normal operation, let's do the upload only when the async operation reach to a terminal state.
Upon user request, let's refer to the other discussion for `velero backup describe --details`